### PR TITLE
Fix concurrent initialization in hadoop compatible FileSystem

### PR DESCRIPTION
**Pull request by cheng-chang**
When HBase gets FileSystem through Hadoop, both Hadoop and HBase will call initialize.

### CODE REVIEW

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 1]
```diff
@@ -455,8 +455,9 @@ public Path getWorkingDirectory() {
   public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
     Preconditions.checkArgument(uri.getScheme().equals(getScheme()),
         PreconditionMessage.URI_SCHEME_MISMATCH.toString(), uri.getScheme(), getScheme());
-    Preconditions.checkArgument(mInitialized.compareAndSet(false, true), "Cannot invoke "
-        + "initialize() more than once");
+    if (!mInitialized.compareAndSet(false, true)) {
```
Comment (gpang) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: If 2 threads initialize(), the first one will succeed, and the second will return. However, If the initialize takes some time, the second thread which returned will not have an initialized client, right?

Comment (cheng-chang) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: This class is not thread safe. But I updated the PR for thread safe concurrent initialization.

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 2]
```diff
@@ -449,14 +450,31 @@ public Path getWorkingDirectory() {
...
     Preconditions.checkArgument(uri.getScheme().equals(getScheme()),
         PreconditionMessage.URI_SCHEME_MISMATCH.toString(), uri.getScheme(), getScheme());
-    Preconditions.checkArgument(mInitialized.compareAndSet(false, true), "Cannot invoke "
-        + "initialize() more than once");
+
+    if (mFileSystem != null) {
+      return;
+    }
+
+    if (mInitializationLock.writeLock().tryLock()) {
```
Comment (gpang) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: can we just simplify this and put this initialization in a `synchronized` block of an object?

Comment (cheng-chang) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: considering that there shouldn't be lots of contentions for initializing, yes.

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 3]
```diff
@@ -91,9 +92,9 @@
   private static final int BLOCK_REPLICATION_CONSTANT = 3;
 
   /** Flag for if the client has been initialized. */
-  private final AtomicBoolean mInitialized = new AtomicBoolean(false);
+  private final ReadWriteLock mInitializationLock = new ReentrantReadWriteLock();
```
Comment (calvinjia) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Is this still used?

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 4]
```diff
@@ -91,9 +92,9 @@
...
 
-  protected FileSystem mFileSystem = null;
+  protected volatile FileSystem mFileSystem = null;
```
Comment (calvinjia) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Does this need to be volatile?

**Comment by gpang** [comment]
This looks good, after getting rid of the extraneous changes/variables that @calvinjia  mentioned.

**Pull request merged**